File: frontend/src/app/core/services/mock/progress.service.ts
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import type {
  DashboardSummaryDto,
  ProgressReportDto,
  LeaderboardEntryDto,
  GenerateReportRequest,
} from '../../models/lesson-planner.models';

function isoDaysAgo(days: number): string {
  const d = new Date();
  d.setDate(d.getDate() - days);
  return d.toISOString();
}

@Injectable({ providedIn: 'root' })
export class MockProgressService {
  private readonly reports: ProgressReportDto[] = [
    {
      id: 1,
      studentId: 1,
      studentName: 'علی احمدی',
      periodStart: isoDaysAgo(70),
      periodEnd: isoDaysAgo(56),
      overallScore: 58.5,
      attendanceRate: 71.4,
      assignmentCompletionRate: 64.3,
      completedAssignments: 9,
      totalAssignments: 14,
      coachNote: 'پیشرفت قابل قبول در پایهٔ ریاضی؛ نیاز به تمرین بیشتر در حل مسائل.',
      generatedAt: isoDaysAgo(55),
      metrics: [
        { id: 1, metricKey: 'math', metricLabel: 'ریاضی', score: 62, target: 100, rank: 1, notes: 'میانگین نمرات تمرین‌ها' },
        { id: 2, metricKey: 'quran', metricLabel: 'قرآن', score: 70, target: 100, rank: 2, notes: 'پیشرفت قرائت و حفظ' },
        { id: 3, metricKey: 'assignments', metricLabel: 'تکالیف', score: 64.3, target: 100, rank: 3, notes: 'تکمیل ۹ از ۱۴ تکلیف' },
        { id: 4, metricKey: 'behavior', metricLabel: 'رفتار', score: 85, target: 100, rank: 4, notes: 'ارزیابی مربی' },
      ],
    },
    {
      id: 2,
      studentId: 1,
      studentName: 'علی احمدی',
      periodStart: isoDaysAgo(56),
      periodEnd: isoDaysAgo(42),
      overallScore: 64.75,
      attendanceRate: 78.6,
      assignmentCompletionRate: 71.4,
      completedAssignments: 10,
      totalAssignments: 14,
      coachNote: 'تکالیف قرآن با نظم بیشتری ارسال شده‌اند.',
      generatedAt: isoDaysAgo(41),
      metrics: [
        { id: 5, metricKey: 'math', metricLabel: 'ریاضی', score: 66, target: 100, rank: 1, notes: 'میانگین نمرات تمرین‌ها' },
        { id: 6, metricKey: 'quran', metricLabel: 'قرآن', score: 76, target: 100, rank: 2, notes: 'پیشرفت قرائت و حفظ' },
        { id: 7, metricKey: 'assignments', metricLabel: 'تکالیف', score: 71.4, target: 100, rank: 3, notes: 'تکمیل ۱۰ از ۱۴ تکلیف' },
        { id: 8, metricKey: 'behavior', metricLabel: 'رفتار', score: 85, target: 100, rank: 4, notes: 'ارزیابی مربی' },
      ],
    },
    {
      id: 3,
      studentId: 1,
      studentName: 'علی احمدی',
      periodStart: isoDaysAgo(42),
      periodEnd: isoDaysAgo(28),
      overallScore: 71.2,
      attendanceRate: 85.7,
      assignmentCompletionRate: 78.6,
      completedAssignments: 11,
      totalAssignments: 14,
      coachNote: 'مشارکت در حلقهٔ حفظ بهبود یافته است.',
      generatedAt: isoDaysAgo(27),
      metrics: [
        { id: 9, metricKey: 'math', metricLabel: 'ریاضی', score: 72, target: 100, rank: 1, notes: 'میانگین نمرات تمرین‌ها' },
        { id: 10, metricKey: 'quran', metricLabel: 'قرآن', score: 81, target: 100, rank: 2, notes: 'پیشرفت قرائت و حفظ' },
        { id: 11, metricKey: 'assignments', metricLabel: 'تکالیف', score: 78.6, target: 100, rank: 3, notes: 'تکمیل ۱۱ از ۱۴ تکلیف' },
        { id: 12, metricKey: 'behavior', metricLabel: 'رفتار', score: 85, target: 100, rank: 4, notes: 'ارزیابی مربی' },
      ],
    },
    {
      id: 4,
      studentId: 1,
      studentName: 'علی احمدی',
      periodStart: isoDaysAgo(28),
      periodEnd: isoDaysAgo(14),
      overallScore: 78.8,
      attendanceRate: 92.9,
      assignmentCompletionRate: 85.7,
      completedAssignments: 12,
      totalAssignments: 14,
      coachNote: 'نتایج آزمون ریاضی به‌طور چشمگیری بهبود یافته است.',
      generatedAt: isoDaysAgo(13),
      metrics: [
        { id: 13, metricKey: 'math', metricLabel: 'ریاضی', score: 79, target: 100, rank: 1, notes: 'میانگین نمرات تمرین‌ها' },
        { id: 14, metricKey: 'quran', metricLabel: 'قرآن', score: 88, target: 100, rank: 2, notes: 'پیشرفت قرائت و حفظ' },
        { id: 15, metricKey: 'assignments', metricLabel: 'تکالیف', score: 85.7, target: 100, rank: 3, notes: 'تکمیل ۱۲ از ۱۴ تکلیف' },
        { id: 16, metricKey: 'behavior', metricLabel: 'رفتار', score: 85, target: 100, rank: 4, notes: 'ارزیابی مربی' },
      ],
    },
    {
      id: 5,
      studentId: 1,
      studentName: 'علی احمدی',
      periodStart: isoDaysAgo(14),
      periodEnd: isoDaysAgo(0),
      overallScore: 85.4,
      attendanceRate: 100,
      assignmentCompletionRate: 92.9,
      completedAssignments: 13,
      totalAssignments: 14,
      coachNote: 'حضور منظم و ارسال به‌موقع همهٔ تکالیف دورهٔ جاری.',
      generatedAt: isoDaysAgo(0),
      metrics: [
        { id: 17, metricKey: 'math', metricLabel: 'ریاضی', score: 86, target: 100, rank: 1, notes: 'میانگین نمرات تمرین‌ها' },
        { id: 18, metricKey: 'quran', metricLabel: 'قرآن', score: 93, target: 100, rank: 2, notes: 'پیشرفت قرائت و حفظ' },
        { id: 19, metricKey: 'assignments', metricLabel: 'تکالیف', score: 92.9, target: 100, rank: 3, notes: 'تکمیل ۱۳ از ۱۴ تکلیف' },
        { id: 20, metricKey: 'behavior', metricLabel: 'رفتار', score: 85, target: 100, rank: 4, notes: 'ارزیابی مربی' },
      ],
    },
  ];

  private readonly leaderboard: LeaderboardEntryDto[] = [
    { studentId: 1, studentName: 'علی احمدی', overallScore: 85.4, rank: 1 },
    { studentId: 2, studentName: 'فاطمه محمدی', overallScore: 82.1, rank: 2 },
    { studentId: 3, studentName: 'محمد رضایی', overallScore: 78.9, rank: 3 },
    { studentId: 4, studentName: 'زهرا کریمی', overallScore: 74.3, rank: 4 },
    { studentId: 5, studentName: 'حسین موسوی', overallScore: 69.7, rank: 5 },
  ];

  getDashboardSummary(): Observable<DashboardSummaryDto> {
    const trend = this.reports
      .slice()
      .sort((a, b) => a.periodEnd.localeCompare(b.periodEnd))
      .map(r => ({
        date: r.periodEnd,
        overallScore: r.overallScore,
        attendanceRate: r.attendanceRate,
        assignmentCompletionRate: r.assignmentCompletionRate,
      }));
    return of({
      recentReports: this.reports.slice().sort((a, b) => b.periodEnd.localeCompare(a.periodEnd)).slice(0, 3),
      trend,
      pendingEvaluations: 2,
      unreadMessages: 3,
      openWorkflows: 1,
      totalXp: 2450,
      leaderboard: this.leaderboard,
    });
  }

  getStudentReports(studentId: number, limit = 12): Observable<ProgressReportDto[]> {
    const filtered = this.reports
      .filter(r => r.studentId === studentId)
      .sort((a, b) => b.periodEnd.localeCompare(a.periodEnd))
      .slice(0, limit);
    return of(filtered);
  }

  generateReport(req: GenerateReportRequest): Observable<ProgressReportDto> {
    const metrics = [
      { id: this.nextMetricId(), metricKey: 'math', metricLabel: 'ریاضی', score: 80, target: 100, rank: 1, notes: 'میانگین نمرات تمرین‌ها' },
      { id: this.nextMetricId(), metricKey: 'quran', metricLabel: 'قرآن', score: 85, target: 100, rank: 2, notes: 'پیشرفت قرائت و حفظ' },
      { id: this.nextMetricId(), metricKey: 'assignments', metricLabel: 'تکالیف', score: 90, target: 100, rank: 3, notes: 'تکمیل تکالیف دوره' },
      { id: this.nextMetricId(), metricKey: 'behavior', metricLabel: 'رفتار', score: 85, target: 100, rank: 4, notes: 'ارزیابی مربی' },
    ];
    const report: ProgressReportDto = {
      id: this.nextReportId(),
      studentId: req.studentId,
      studentName: this.reports.find(r => r.studentId === req.studentId)?.studentName ?? `دانش‌آموز ${req.studentId}`,
      periodStart: req.periodStart,
      periodEnd: req.periodEnd,
      overallScore: Math.round((metrics.reduce((sum, m) => sum + m.score, 0) / metrics.length) * 100) / 100,
      attendanceRate: 90,
      assignmentCompletionRate: 90,
      completedAssignments: 12,
      totalAssignments: 14,
      coachNote: req.coachNote,
      generatedAt: new Date().toISOString(),
      metrics,
    };
    this.reports.push(report);
    return of(report);
  }

  getLeaderboard(limit = 10): Observable<LeaderboardEntryDto[]> {
    return of(this.leaderboard.slice(0, limit));
  }

  private nextReportId(): number {
    return Math.max(0, ...this.reports.map(r => r.id)) + 1;
  }

  private nextMetricId(): number {
    return Math.max(0, ...this.reports.flatMap(r => r.metrics.map(m => m.id))) + 1;
  }
}

File: backend/src/EducationalPlatform.Nehzat.Infrastructure/Services/ProgressService.cs
using EducationalPlatform.Nehzat.Application.DTOs;
using EducationalPlatform.Nehzat.Application.Interfaces;
using EducationalPlatform.Nehzat.Domain.Entities;
using EducationalPlatform.Nehzat.Domain.Entities.Math;
using EducationalPlatform.Nehzat.Domain.Entities.Quran;
using EducationalPlatform.Nehzat.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace EducationalPlatform.Nehzat.Infrastructure.Services;

public class ProgressService : IProgressService
{
    private readonly AppDbContext _db;

    public ProgressService(AppDbContext db) => _db = db;

    public async Task<DashboardSummaryDto> GetDashboardAsync(int userId)
    {
        var user = await _db.Users.FindAsync(userId);
        var isStudent = user?.StudentId != null;

        List<ProgressReport> reports;
        if (isStudent && user!.StudentId != null)
        {
            reports = await _db.ProgressReports
                .Include(r => r.Metrics)
                .Include(r => r.Student)
                .Where(r => r.StudentId == user.StudentId)
                .OrderByDescending(r => r.PeriodEnd)
                .Take(6)
                .ToListAsync();
        }
        else
        {
            var all = await _db.ProgressReports
                .Include(r => r.Metrics)
                .Include(r => r.Student)
                .OrderByDescending(r => r.PeriodEnd)
                .ToListAsync();
            reports = all.GroupBy(r => r.StudentId)
                .Select(g => g.First())
                .Take(6)
                .ToList();
        }

        var recentReports = reports.Take(3).Select(r => ToDto(r)).ToList();
        var trend = reports
            .OrderBy(r => r.PeriodEnd)
            .Select(r => new ProgressTrendPointDto
            {
                Date = r.PeriodEnd,
                OverallScore = r.OverallScore,
                AttendanceRate = r.AttendanceRate,
                AssignmentCompletionRate = r.AssignmentCompletionRate
            })
            .ToList();

        var unreadMessages = await CountUnreadMessagesAsync(userId);
        var openWorkflows = await _db.WorkflowRequests.CountAsync(w => w.Status == "pending");
        var totalXp = await _db.UserXp
            .Where(x => x.UserId == userId)
            .Select(x => (int?)x.TotalXp)
            .FirstOrDefaultAsync() ?? 0;

        var pendingEvaluations = isStudent
            ? await _db.AssessmentResults.CountAsync(a => a.StudentId == user!.StudentId && a.Status == "pending")
            : await _db.AssessmentResults.CountAsync(a => a.Status == "pending");

        return new DashboardSummaryDto
        {
            RecentReports = recentReports,
            Trend = trend,
            PendingEvaluations = pendingEvaluations,
            UnreadMessages = unreadMessages,
            OpenWorkflows = openWorkflows,
            TotalXp = totalXp,
            Leaderboard = await GetLeaderboardAsync(userId, 5)
        };
    }

    public async Task<List<ProgressReportDto>> GetReportsByStudentAsync(int userId, int studentId, int limit = 12)
    {
        var reports = await _db.ProgressReports
            .Include(r => r.Metrics)
            .Include(r => r.Student)
            .Where(r => r.StudentId == studentId)
            .OrderByDescending(r => r.PeriodEnd)
            .Take(limit)
            .ToListAsync();
        return reports.Select(ToDto).ToList();
    }

    public async Task<ProgressReportDto> GenerateReportAsync(int userId, GenerateReportRequest req)
    {
        var mathScore = await AverageOrZeroAsync(
            _db.MathProgresses.Where(p => p.StudentId == req.StudentId && p.Score.HasValue).Select(p => (decimal?)p.Score));

        var quranScore = await AverageOrZeroAsync(
            _db.QuranStudentProgresses.Where(q => q.StudentId == req.StudentId).Select(q => (decimal?)q.Percentage));

        var submissions = await _db.AssignmentSubmissions
            .Where(s => s.StudentId == req.StudentId
                && s.SubmissionDate >= req.PeriodStart.Date
                && s.SubmissionDate <= req.PeriodEnd.Date)
            .ToListAsync();
        var completed = submissions.Count(s => s.IsCompleted);
        var total = submissions.Count;
        var completionRate = total > 0 ? Math.Round((decimal)completed / total * 100m, 2) : 0m;

        var metrics = new List<ProgressMetric>
        {
            new()
            {
                MetricKey = "math",
                MetricLabel = "ریاضی",
                Score = mathScore,
                Target = 100,
                Rank = 1,
                Notes = "میانگین نمرات تمرین‌های ریاضی"
            },
            new()
            {
                MetricKey = "quran",
                MetricLabel = "قرآن",
                Score = quranScore,
                Target = 100,
                Rank = 2,
                Notes = "میانگین درصد پیشرفت قرائت و حفظ"
            },
            new()
            {
                MetricKey = "assignments",
                MetricLabel = "تکالیف",
                Score = completionRate,
                Target = 100,
                Rank = 3,
                Notes = $"تکمیل {completed} از {total} تکلیف دوره"
            },
            new()
            {
                MetricKey = "behavior",
                MetricLabel = "رفتار",
                Score = 85,
                Target = 100,
                Rank = 4,
                Notes = "ارزیابی مربی"
            }
        };

        var overall = Math.Round(metrics.Average(m => m.Score), 2);

        var report = new ProgressReport
        {
            StudentId = req.StudentId,
            PeriodStart = req.PeriodStart,
            PeriodEnd = req.PeriodEnd,
            OverallScore = overall,
            AttendanceRate = completionRate,
            AssignmentCompletionRate = completionRate,
            CompletedAssignments = completed,
            TotalAssignments = total,
            CoachNote = req.CoachNote,
            GeneratedAt = DateTime.UtcNow,
            Metrics = metrics
        };

        _db.ProgressReports.Add(report);
        await _db.SaveChangesAsync();
        return ToDto(report);
    }

    public async Task<List<ProgressLeaderboardEntryDto>> GetLeaderboardAsync(int userId, int limit = 10)
    {
        var all = await _db.ProgressReports
            .Include(r => r.Student)
            .OrderByDescending(r => r.PeriodEnd)
            .ToListAsync();

        var top = all.GroupBy(r => r.StudentId)
            .Select(g => g.First())
            .OrderByDescending(r => r.OverallScore)
            .Take(limit)
            .ToList();

        return top
            .Select((r, i) => new ProgressLeaderboardEntryDto
            {
                StudentId = r.StudentId,
                StudentName = r.Student != null
                    ? $"{r.Student.FirstName} {r.Student.LastName}".Trim()
                    : $"دانش‌آموز {r.StudentId}",
                OverallScore = r.OverallScore,
                Rank = i + 1
            })
            .ToList();
    }

    private async Task<int> CountUnreadMessagesAsync(int userId)
    {
        var memberships = await _db.ConversationMembers
            .Where(cm => cm.UserId == userId)
            .Select(cm => new { cm.ConversationId, cm.LastReadAt })
            .ToListAsync();

        if (memberships.Count == 0) return 0;

        var conversationIds = memberships.Select(m => m.ConversationId).ToList();

        var messages = await _db.Messages
            .Where(m => conversationIds.Contains(m.ConversationId)
                && m.SenderId != userId
                && !m.IsDeleted)
            .Select(m => new { m.ConversationId, m.CreatedAt })
            .ToListAsync();

        var unread = 0;
        foreach (var membership in memberships)
        {
            var lastReadAt = membership.LastReadAt;
            unread += messages.Count(m =>
                m.ConversationId == membership.ConversationId &&
                (!lastReadAt.HasValue || m.CreatedAt > lastReadAt.Value));
        }
        return unread;
    }

    private static async Task<decimal> AverageOrZeroAsync(IQueryable<decimal?> query)
    {
        var values = await query.ToListAsync();
        if (values.Count == 0) return 0m;
        var average = values.Where(v => v.HasValue).Select(v => v!.Value).DefaultIfEmpty(0m).Average();
        return Math.Clamp(average, 0m, 100m);
    }

    private static ProgressReportDto ToDto(ProgressReport r) => new()
    {
        Id = r.Id,
        StudentId = r.StudentId,
        StudentName = r.Student != null
            ? $"{r.Student.FirstName} {r.Student.LastName}".Trim()
            : $"دانش‌آموز {r.StudentId}",
        PeriodStart = r.PeriodStart,
        PeriodEnd = r.PeriodEnd,
        OverallScore = r.OverallScore,
        AttendanceRate = r.AttendanceRate,
        AssignmentCompletionRate = r.AssignmentCompletionRate,
        CompletedAssignments = r.CompletedAssignments,
        TotalAssignments = r.TotalAssignments,
        CoachNote = r.CoachNote,
        GeneratedAt = r.GeneratedAt,
        Metrics = r.Metrics.OrderBy(m => m.Rank).Select(m => new ProgressMetricDto
        {
            Id = m.Id,
            MetricKey = m.MetricKey,
            MetricLabel = m.MetricLabel,
            Score = m.Score,
            Target = m.Target,
            Rank = m.Rank,
            Notes = m.Notes
        }).ToList()
    };
}
